[NO-TICKET] Update WRITING-DOCUMENTATION.md (#739)
CMSgov/design-system: guides/WRITING-DOCUMENTATION.md

- [Folder structure](#folder-structure)
- [Writing component documentation](#writing-component-documentation)
- [Writing general documentaion](#writing-general-documentation)

# Folder structure

### The docs folder structure

**Note:** Name and structure components as follows in order for the documentation site generation to work as expected.

For a component named "Foo", you would have the following:

```
docs/src
└── pages
    ├── components
    |   └── Foo                     // Folder for component documentation content and code examples
    |       ├── Foo.docs.scss       // Component documentation wirtten in KSS
    |       ├── Foo.example.jsx     // React component example
    |       └── Foo.example.html    // HTML example
    ├── guidelines      // Folder for documentation guideline pages
    ├── patterns        // Folder for pattern documentation and example files
    ├── startup         // Folder for documentation statup pages
    ├── styles          // Folder for styles for the documentation site
    └── utilities       // Folder for Utility documentation and example files
```

# Writing component documentation

> [KSS](https://github.com/kss-node/kss-node) is the primary way documentation is written.

Reviewer: ```
[KSS](https://github.com/kss-node/kss-node) is the primary way documentation is written for pages with code examples, i.e. components, patterns, utilities.
```

Reviewer: ```suggestion
[KSS](https://github.com/kss-node/kss-node) is the primary way documentation is written for pages with code examples, i.e. components, patterns, utilities.
```

Documentation for the design system's components are automatically generated from their CSS and JS comments. In order for this to work, you need to follow some conventions set by the tools we use to generate the documentation: [KSS](https://github.com/kss-node/kss-node) and [react-docgen](https://github.com/reactjs/react-docgen).

- [Folder structure](#folder-structure)
- [Writing component documentation](#writing-component-documentation)
- [Writing general documentaion](#writing-general-documentation)

# Folder structure

### The docs folder structure

> **Note:** Name and structure components as follows in order for the documentation site generation to work as expected.

Reviewer: ```suggestion
**Note:** Documentation files must be named and organized with the following structure  in order for the documentation site generation to work as expected.
```

Documentation for the design system's components are automatically generated from their CSS and JS comments. In order for this to work, you need to follow some conventions set by the tools we use to generate the documentation: [KSS](https://github.com/kss-node/kss-node) and [react-docgen](https://github.com/reactjs/react-docgen).

- [Directory structure](#directory-structure)
  - [The docs folder structure](#the-docs-folder-structure)
  - [The src folder structure](#the-src-folder-structure)

> - [Component documentation page formatting](#component-documentation-page-formatting)

Reviewer: I'm reading through this page structure and it feels like it could be clarified a bit. "Component documentation page formatting" vs "Documenting React components" sounds a bit confusing to me. I wonder if we can reorganize this document on a high level to be something like this:

```
- Folder structure
- Writing component documentation
   - Intro to KSS, when to use it
   - flags, markup, styleguide, etc
   - example
- Writing general documentation
   - front-matter
   - properties
   - example
```

  - [Heading (required)](heading-required)
  - [Description](#description)
  - [@Flags](#flags)
  - [Modifiers](#modifiers)
  - [Markup](#markup)
  - [Style guide](#style-guide)
  - [Guidance](#guidance)
- [Sample component documentation page](#sample-component-documentation-page)
- [Documenting React components](#documenting-react-components)
  - [Description](#description-1)
  - [PropTypes](#proptypes)
- [General content page](#general-content-page)
  - [Front-matter](#front-matter)
  - [Page content](#page-content)

## Directory structure

> ```

Reviewer 2: Can we remove `child` as the directory structure applies for both core & child ds?

Reviewer: I would even be ok with just removing this beginning folder structure. I think it's enough to have a major section that says `Docs folder structure`, we probably dont need to go into the design system package source folder

child design system root
├── `docs`    Directory where documentation and code examples are located
└── `src`     Directory where component code, styles, fonts, and images are located
```

### The docs folder structure

**Note:** Name and structure components as follows in order for the documentation site generation to work as expected.

For a component named "Foo", you would have the following:

```
child design system root

>   └── docs

Reviewer: This can also be simplified to `docs/src` instead of double nesting

      └── src
          └── pages

>               └── components

Reviewer: We should include the other major page sections

                  └── Foo                     Directory for component documentation content and code examples
                      ├── Foo.docs.scss       Component documentation wirtten in KSS
                      ├── Foo.example.jsx     React component example
                      └── Foo.example.html    HTML example
```

> ### The src folder structure

Reviewer: I dont think this is necessary to include this section, as it doesnt have to do with writing documentation

```
└── components

>     └── Foo                     A directory holding all relevant code

Reviewer: I feel like this description isn’t super helpful. It’s a directory holding documentation content and examples, we don’t want to confuse it with the source code

Reviewer: Another though, we should probably show how the documentation files are separated from the source code. Maybe have a separate section that shows the file structure of ‘docs/src/pages/components’?

To extend the default functionality of KSS, we've implemented support for custom flags. Flags should always come after the description.

> | flag name                   | example                                                      | flag description                                                                                                                        |

Reviewer: Great idea for the table!

| **`@react-props [NAME]`**   | `@react-props Button.jsx`                                    | Displays the React prop documentation for a component. `[NAME]` is the filename of the react component where the Proptypes are defined. |
| **`@react-example [NAME]`** | `@react-example Button.example.jsx`                          | Displays the example file using this React component. `[NAME]` is the filename of the react example to be displayed.                    |
| **`@responsive`**           | `@responsive`                                                | Renders breakpoint toggles for the markup example.                                                                                      |
| **`@status [NAME]`**        | `@status Draft`                                              | Displays a status badge. `[NAME]` must be one of the following values: `Draft`, `Work in progress`, `Ready`, `Deprecated`.              |
| **`@uswds [URL]`**          | `@uswds https://designsystem.digital.gov/components/button/` | `[URL]` is the link to the corresponding US Web Design System component.                                                                |

### Modifiers

> 

Reviewer: I believe we are in the process of deprecating modifiers. There are still pages that have it and our code still supports it, but maybe we should remove them from the documentation?

Reviewer: Actually, we can remove it from the docs after its deprecated

Author: Yeah, I left it because it's still in use. I was thinking it could be worth revisiting as it's pretty powerful to show iterations of things without having to write a bunch of duplicate code.

Reviewer: My understanding is that we moved away from that modifier model when we started to consolidate examples like with the button component. Modifiers can be convenient, but it also assumes the maintainer knows EJS (the templating language needed to use modifiers) and knows how the modifier generation works.

If we are eventually moving to markdown, this sort of organization wouldnt work as well long term, and i think its better to be explicit

Author: Yeah, that's a good point - I think we should remove the modifiers section from the docs then if we don't want folks using them.

If the item you are documenting has multiple states or styles depending on added classes or pseudo-classes, you should document them in the modifiers section. [More info on modifiers from the KSS documentation site](https://warpspire.com/kss/syntax/).

### Markup

> Markup examples shold be written in a separate plain `.html` file, relative to the SCSS file:

Reviewer: ```suggestion
Markup examples should be written in a separate `example.html` file, which should be located in the same directory as the SCSS docs file:
```

```
Markup: Button.example.html
```

### Style guide

> The `reference` defines the documentation site's structure.

Reviewer: ```suggestion
The `reference` defines the documentation page's position or URL within the site structure.
```

| **`@status [NAME]`**        | `@status Draft`                                              | Displays a status badge. `[NAME]` must be one of the following values: `Draft`, `Work in progress`, `Ready`, `Deprecated`.              |
| **`@uswds [URL]`**          | `@uswds https://designsystem.digital.gov/components/button/` | `[URL]` is the link to the corresponding US Web Design System component.                                                                |

### Modifiers

If the item you are documenting has multiple states or styles depending on added classes or pseudo-classes, you should document them in the modifiers section. [More info on modifiers from the KSS documentation site](https://warpspire.com/kss/syntax/).

### Markup

Markup examples should be written in a separate `example.html` file, which should be located in the same directory as the SCSS docs file:

```
Markup: Button.example.html
```

### Style guide

The `reference` defines the position or URL within the site structure for a section of KSS documentation.

The expected format is: `[Top-level slug].[Subpage slug].[Page section slug]`.

> For example, `Style guide: components.button.disabled`.

Reviewer: ```suggestion
For example:
```
Style guide: components.button.disabled
```
```

The docs site supports a maximum of 2 levels of nesting. Pages nested a 3rd level will be displayed in the body of their parent page.

Pages generated from KSS comment blocks are ordered alphabetically, and page sections are displayed in the order in which they are in the SCSS file.

### Guidance

To add a guidance section to the page the `style guide:` `[Page section slug]` should be `guidance`.

> For example: `Style guide: components.button.guidance`

Reviewer: Same here, I find the code block helps organize this a bit.

For example:
```
Style guide: components.button.guidance
```

Guidance should be written using [Markdown](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet) if you include `---` right after your opening section comments, `/*`.

> To link to another documentation page in your guidance content use a relative url and the `{{root}}` template. This template will be replaced with your site's root url.

Reviewer: Can we remove line 111 and put the example in the sentence. This way we can avoid using a code block for the example

```suggestion
To link to another documentation page in your guidance content use a relative url and the `{{root}}` template (i.e. `{{root}}/components/button`). This template will be replaced with your site's root url.
```

For example, `{{root}}/components/button`

Documentation and guidance should be placed at the end of the SCSS file and conform to the following format and sections:

<details>
<summary><b>View guidance section example</b></summary>

```scss
/*
---

### When to use

List typical use cases for this component.

### When to consider alternatives

List commonly confused use cases where an alternative component would be a better solution.

### Usage

List any design or content guidelines related to how this component should be used.

### Accessibility

List any accessibility considerations like required keyboard functionality, ARIA attributes, etc...

### Theming (optional)

List any Sass variables a developer could override to customize the appearance of the component.

### Related patterns

List any related components or utility classes that connect with the component, whether as a parent, sibling, or child within the pattern's taxonomy.

### Learn more 

List any links that served as inspiration, references, or research related to this component.

Style guide: components.component-name.guidance
*/
```

</details>

## Documenting React components

> Using [`react-docgen`](https://github.com/reactjs/react-docgen), we extract a description for each of the component's `propTypes` from comments in React component files.

Reviewer: Maybe we can move this `react-docgen` info under the Proptypes section

> ### Description

Reviewer: Can we remove this heading? Description doesnt really mean anything in the context of this section

To extend the default functionality of KSS, we've implemented support for custom flags. Flags should always come after the description.

| flag name                   | example                                                      | flag description                                                                                                                        |
| --------------------------- | ------------------------------------------------------------ | --------------------------------------------------------------------------------------------------------------------------------------- |

> | **`@hide-markup`**          | `@hide markup`                                               | Hides the code snippet.                                                                                                                 |

Reviewer 2: missing `-` for `@hide-markup`

### Markup

Markup examples shold be written in a separate plain `.html` file, relative to the SCSS file:

```
Markup: Button.example.html
```

### Style guide

The `reference` defines the documentation site's structure.

The expected format is: `[Top-level slug].[Subpage slug].[Page section slug]`.
For example, `Style guide: components.button.disabled`.

The docs site supports a maximum of 2 levels of nesting. Pages nested a 3rd level will be displayed in the body of their parent page.

Pages generated from KSS comment blocks are ordered alphabetically, and page sections are displayed in the order in which they are in the SCSS file.

### Guidance

To add a guidance section to the page the `style guide:` `[Page section slug]` should be `guidance`.
For example: `Style guide: components.button.guidance`

Guidance should be written using [Markdown](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet) if you include `---` right after your opening section comments, `/*`.

> To link to another documentation page in your guidance content use a relative url.

Reviewer: ```suggestion
To link to another documentation page in your guidance content use a relative url and the `{{root}}` template. This template will replaced with your site's root url.
```

> For example, `/components/button`

Reviewer: ```suggestion
For example, `{{root}}/components/button`
```

> To link to another documentation page in your guidance content use a relative url and the `{{root}}` template. This template will replaced with your site's root url.

Reviewer 2: missing `be`? will `be` replaced with your site's root url.

Documentation and guidance should be placed at the end of the SCSS file and conform to the following format and sections:

<details>

> <summary><b>View guidance section example</b></summary>

Reviewer: Where did these come from?

Author: The guidance section was not in a collapsable details view previously - see here - https://github.com/CMSgov/design-system/pull/739/files#diff-36dd66eb15347ca0028585e9135a481cL126

Reviewer: Ah I see, I never saw that collapseable view before.

```scss
/*
---

### When to use

List typical use cases for this component.

### When to consider alternatives

List commonly confused use cases where an alternative component would be a better solution.

### Usage

List any design or content guidelines related to how this component should be used.

### Accessibility

List any accessibility considerations like required keyboard functionality, ARIA attributes, etc...

### Theming (optional)

List any Sass variables a developer could override to customize the appearance of the component.

### Related patterns

List any related components or utility classes that connect with the component, whether as a parent, sibling, or child within the pattern's taxonomy.

### Learn more 

List any links that served as inspiration, references, or research related to this component.

Style guide: components.component-name.guidance
*/
```

</details>

## Example

> The example below is a documentation page with a URL path of `/components/buttons`.

Reviewer: Could we just link to the button docs file directly instead of copying most of it here?

Reviewer: Also, I feel this example should be at the end of this section, after we explain how to document react

Author: Yes on linking to the button docs file 👍 
Yes on moving to the bottom of this section 👍

## Example

> Tkae a look at the [button component documentation (_Button.docs.scss)](https://github.com/CMSgov/design-system/blob/master/packages/design-system-docs/src/pages/components/Button/_Button.docs.scss) for an example.

Reviewer 2: Typo `Take`

```
---
label: Hello World
title: Hello world
markup: hello-world.example.html
weight: 100
responsive: true
uswds: https://designsystem.digital.gov/page-templates/
usage: |

>   Hellow world this is the description of the page

Reviewer: ```suggestion
  Hello world this is the description of the page
```
